packages/state-transition/src/block/processExecutionLayerWithdrawalRequest.ts

```diff
@@ -0,0 +1,103 @@
+import {computeExitEpochAndUpdateChurn} from "../util/epoch.js";
+import {initiateValidatorExit} from "./initiateValidatorExit.js";
+
+export function processExecutionLayerWithdrawalRequest(
+  state: CachedBeaconStateElectra,
+  executionLayerWithdrawalRequest: electra.ExecutionLayerWithdrawalRequest
+): void {
+  const amount = Number(executionLayerWithdrawalRequest.amount);
+  const {pendingPartialWithdrawals, validators, epochCtx} = state;
+  const {pubkey2index, config} = epochCtx; // TODO Electra: Use finalized+unfinalized pubkey cache from 6110
+  const isFullExitRequest = amount === FULL_EXIT_REQUEST_AMOUNT;
+
+  // If partial withdrawal queue is full, only full exits are processed
+  if (pendingPartialWithdrawals.length >= PENDING_PARTIAL_WITHDRAWALS_LIMIT && !isFullExitRequest) {
+    return;
+  }
+
+  const validatorIndex = pubkey2index.get(executionLayerWithdrawalRequest.validatorPubkey);
+
+  if (validatorIndex === undefined) {
+    throw new Error(
+      `Can't find validator index from ExecutionLayerWithdrawalRequest : pubkey=${toHexString(
+        executionLayerWithdrawalRequest.validatorPubkey
+      )}`
+    );
+  }
+
+  const validator = validators.getReadonly(validatorIndex);
+
+  if (!isValidValidator(validator, executionLayerWithdrawalRequest.sourceAddress, state)) {
+    return;
+  }
+
+  // TODO Electra: Consider caching pendingPartialWithdrawals
+  const pendingBalanceToWithdraw = getPendingBalanceToWithdraw(state, validatorIndex);
+  const validatorBalance = state.balances.get(validatorIndex);
+
+  if (isFullExitRequest) {
+    // only exit validator if it has no pending withdrawals in the queue
+    if (pendingBalanceToWithdraw === 0) {
+      initiateValidatorExit(ForkSeq.electra, state, validator);
+    } else {
+      // Full request can't be fulfilled because still has pending withdrawals.
+      // TODO Electra: add log here
+    }
+    return;
+  }
+
+  const hasSufficientEffectiveBalance = validator.effectiveBalance >= MIN_ACTIVATION_BALANCE;
+  const hasExcessBalance = validatorBalance > MIN_ACTIVATION_BALANCE + pendingBalanceToWithdraw;
+
+  // Only allow partial withdrawals with compounding withdrawal credentials
+  if (
+    hasCompoundingWithdrawalCredential(validator.withdrawalCredentials) &&
+    hasSufficientEffectiveBalance &&
+    hasExcessBalance
+  ) {
+    const amountToWithdraw = BigInt(
+      Math.min(validatorBalance - MIN_ACTIVATION_BALANCE - pendingBalanceToWithdraw, amount)
+    );
+    const exitQueueEpoch = computeExitEpochAndUpdateChurn(state, amountToWithdraw);
+    const withdrawableEpoch = exitQueueEpoch + config.MIN_VALIDATOR_WITHDRAWABILITY_DELAY;
+
+    const pendingPartialWithdrawal = ssz.electra.PartialWithdrawal.toViewDU({
+      index: validatorIndex,
+      amount: amountToWithdraw,
+      withdrawableEpoch,
+    });
+    state.pendingPartialWithdrawals.push(pendingPartialWithdrawal);
+  }
+}
+
+function isValidValidator(
```

Comment:
this function name is confusing, maybe `isValidatorEligibleForWithdraw()` or something else?